### DISCUSSION TITLE
Zoom/Scroll aware Grid

### DISCUSSION THREAD

**Discussion by frankbenoit** [imixs/open-bpmn#316] [Ideas]
Hi

i come across Open-BPMN as a GLSP example that we (me and my collegue) can take inspiration for our project.
I noticed that the visible grid in Open-BPMN is not yet aware of zooming and scroll.
We solved this with 
````
    configureModelElement(context, DefaultTypes.GRAPH, GGraph, MyGraphView);

@injectable()
export class MyGraphView implements IView {
    @inject(EdgeRouterRegistry) edgeRouterRegistry: EdgeRouterRegistry;

    render(model: Readonly<SGraphImpl>, context: RenderingContext): VNode {
        const edgeRouting = this.edgeRouterRegistry.routeAllChildren(model);

        const gridBounds: Bounds = {
            x: model.canvasBounds.x / model.zoom + model.scroll.x,
            y: model.canvasBounds.y / model.zoom + model.scroll.y,
            width: Math.max(10, model.canvasBounds.width / model.zoom),
            height: Math.max(10, model.canvasBounds.height / model.zoom)
        };

        const transform = `scale(${model.zoom}) translate(${-model.scroll.x},${-model.scroll.y})`;
        return (
            <svg class-sprotty-graph={true}>
                <defs>
                    <pattern id='my-grid-pattern' x='-5' y='-5' width='10' height='10' patternUnits='userSpaceOnUse'>
                        <circle id='pattern-circle' cx='5' cy='5' r='0.5' class-grid-dot={true} />
                    </pattern>
                </defs>

                <g transform={transform}>
                    <rect
                        x={gridBounds.x}
                        y={gridBounds.y}
                        width={gridBounds.width}
                        height={gridBounds.height}
                        fill='url(#my-grid-pattern)'
                    />
                    <g class-graph-content={true}>{context.renderChildren(model, { edgeRouting })}</g>
                </g>
            </svg>
        );
    }
}
````

**Comment by rsoika** [comment]
Hi @frankbenoit , that's interesting and thanks for your contribution!

But how did you solve the dimensions problem of the diagram plane. In my test the grid is not filling the visible part of the plane. 

![Peek 2023-12-27 10-00](https://github.com/imixs/open-bpmn/assets/1339124/54924848-ca74-4242-abd9-6d843bf214d5)

I had a similar problem with my helper-line feature. 

![Peek 2023-12-27 10-04](https://github.com/imixs/open-bpmn/assets/1339124/36ee3dc1-bfee-46ae-b7c8-b8740845a384)

Also here my helper lines are sometimes not filling the complete plane and are starting somewhere in the middle of the diagram.

**Comment by frankbenoit** [comment]
I think there is just an incorrect offset in the rect with the fill. The grid seems to be fine and the elements align with the dots. Just the x/y of the rect which limits the pattern shown seems to be wrong coordinates.

**Comment by rsoika** [comment]
Yes I think the start position should be the `model.scroll.x/y` value. So my implementation looks now like this:

```javascript
@injectable()
export class BPMNGraphView implements IView {
    @inject(EdgeRouterRegistry) edgeRouterRegistry: EdgeRouterRegistry;

    render(model: Readonly<SGraphImpl>, context: RenderingContext): VNode {
        const edgeRouting = this.edgeRouterRegistry.routeAllChildren(model);

        const gridBounds: Bounds = {
            x: model.scroll.x,
            y: model.scroll.y,
            width: Math.max(10, model.canvasBounds.width / model.zoom),
            height: Math.max(10, model.canvasBounds.height / model.zoom)
        };

        const transform = `scale(${model.zoom}) translate(${-model.scroll.x},${-model.scroll.y})`;
        return (
            <svg class-sprotty-graph={true}>
                <defs>
                    <pattern id='bpmn-grid-pattern' x='-5' y='-5' width='10' height='10' patternUnits='userSpaceOnUse'>
                        <line x1="0" y1="5" x2="10" y2="5" class-bpmn-grid-line={true} />
                        <line x1="5" y1="0" x2="5" y2="10" class-bpmn-grid-line={true} />
                    </pattern>
                </defs>
                <g transform={transform}>
                    <rect
                        x={gridBounds.x}
                        y={gridBounds.y}
                        width={gridBounds.width}
                        height={gridBounds.height}
                        fill='url(#bpmn-grid-pattern)'
                    />
                    <g class-graph-content={true}>{context.renderChildren(model, { edgeRouting })}</g>
                </g>
            </svg>
        );
    }
}
```

I am using lines instead of dots.

![Peek 2023-12-27 14-25](https://github.com/imixs/open-bpmn/assets/1339124/89e43970-cd1f-4905-99b6-2ef9104edcea)

In my CSS file I also use the theia/vscode vars for coloring: 

```css
.bpmn-grid-line {
    stroke-width: 0.5;
    stroke: var(--theia-titleBar-activeBackground);
}
```

This adapts the line color to the current theme:

![grafik](https://github.com/imixs/open-bpmn/assets/1339124/ef321129-49a3-4093-a065-6723a6bf7dda)


To configure the view in my Diagram Container I use `GLSPGraph` instead of `SGraph`:

```javascript
configureModelElement(context, DefaultTypes.GRAPH, GLSPGraph, BPMNGraphView);
```

this may depend on the GLSP version you are using... Which GLSP version are you using currently?


And still I have the issue that opening a new diagram the grid is lost for some reason....

**Reply by frankbenoit** [reply]
The offset in your first reply, could it be this one?
![image](https://github.com/imixs/open-bpmn/assets/8962790/35c67963-7f2c-4679-88aa-13c26e642afe)

And because I set the rect in transformed coordinates `model.canvasBounds.x / model.zoom + model.scroll.x` seems to make sense. I tried to make a colored stroke for the rect and shrink it a bit, and it perfectly fits the canvas on every scroll and zoom.

I am new to GLSP, so starting with the current 2.0.

About opening, I see no problem here. But I have to admit, my project runs inside of an Eclipse RCP, so setup might be different here.

**Reply by rsoika** [reply]
Ok, with 2.0 there came a lot of changes that may explain the difference in some of the details. I tested it with 1.x but I currently work on a migration to 2.0. 
Once again - thanks for your idea to implement a zoom aware Grid :+1:
